src/cascadia/TerminalSettingsModel/ActionMap.cpp

```diff
@@ -0,0 +1,557 @@
+    // - Update our internal state with the key chord of the newly registered action
+    // Arguments:
+    // - cmd: the action we're trying to register
+    // - oldCmd: the action that already exists in our internal state. May be null.
+    // - consolidatedCmd: the consolidated action that already exists in our internal state. May be null.
+    void ActionMap::_TryUpdateKeyChord(const Model::Command& cmd, const Model::Command& oldCmd, const Model::Command& consolidatedCmd)
+    {
+        if (const auto keys{ cmd.Keys() })
+        {
+            const auto oldKeyPair{ _KeyMap.find(keys) };
+            if (oldKeyPair != _KeyMap.end())
+            {
+                // Collision: The key chord was already in use.
+                // Remove the old one.
+                const auto& actionPair{ _ActionMap.find(oldKeyPair->second) };
+                const auto& conflictingCmd{ actionPair->second };
+                const auto& conflictingCmdImpl{ get_self<implementation::Command>(conflictingCmd) };
+                conflictingCmdImpl->EraseKey(keys);
+            }
+            else if (const auto& conflictingCmd{ GetActionByKeyChord(keys) })
+            {
+                // Collision: The key chord was already in use, but by an action in another layer
+                const auto conflictingActionID{ ActionHash{}(conflictingCmd.ActionAndArgs()) };
+                const auto& consolidatedCmdPair{ _ConsolidatedActions.find(conflictingActionID) };
+                if (consolidatedCmdPair != _ConsolidatedActions.end())
+                {
+                    const auto& consolidatedCmdImpl{ get_self<implementation::Command>(consolidatedCmdPair->second) };
+                    consolidatedCmdImpl->EraseKey(keys);
+                }
+                else
+                {
+                    // Create a copy of the conflicting action,
+                    // and erase the conflicting key chord from the copy.
+                    const auto& conflictingCmdImpl{ get_self<implementation::Command>(conflictingCmd) };
+                    const auto& conflictingCmdCopy{ conflictingCmdImpl->Copy() };
+                    conflictingCmdCopy->EraseKey(keys);
+                    _ConsolidatedActions.emplace(conflictingActionID, *conflictingCmdCopy);
+                }
+            }
+
+            // Assign the new action.
+            const auto actionID{ ActionHash{}(cmd.ActionAndArgs()) };
+            _KeyMap.insert_or_assign(keys, actionID);
+
+            if (oldCmd)
+            {
+                // Update inner Command with new key chord
+                auto oldCmdImpl{ get_self<Command>(oldCmd) };
+                oldCmdImpl->RegisterKey(keys);
+            }
+
+            if (consolidatedCmd)
+            {
+                // Update inner Command with new key chord
+                auto consolidatedCmdImpl{ get_self<Command>(consolidatedCmd) };
+                consolidatedCmdImpl->RegisterKey(keys);
+            }

```

Comment:
okay so we get here when... let me see if i can figure this out

we are updating a command from a lower layer to have keys that they didn't otherwise have, and we *did not conflict* any of our keys or the parent's keys?

I wonder if there's a clever way to handle the "copy-up and mutate" state with a helper

Reply:
for each of these "Try Update" functions I'd like you to comment what state gets us into each conditional -- so somebody else can follow it later!

Reply:
Added a ton of examples and moved the logic around a bit. Hope this helps make it clearer!